sale_timesheet_limit_date/models/account_invoice_line.py

```diff
@@ -0,0 +1,36 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, models
+from odoo.osv import expression
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+
+    @api.model
+    def create(self, values):
+        # unlink lines from invoice if they under limit_date
+        res = super().create(values)
+        for invoice_line in res:
+            to_clean = self.env['account.analytic.line']
+            if (invoice_line.invoice_id.type == 'out_invoice'
+                    and invoice_line.invoice_id.state == 'draft'):
+                sale_line_delivery = invoice_line.sale_line_ids.filtered(
+                    lambda sol: sol.product_id.invoice_policy == 'delivery'
+                    and sol.product_id.service_type == 'timesheet')
+                if sale_line_delivery:
+                    domain = [(
+                        'timesheet_invoice_id', '=', invoice_line.invoice_id.id
+                    )]
+                    old_timesheets = self.env[
+                        'account.analytic.line'].search(domain)
+                    domain = expression.AND([domain, [
+                        ('date', '<=',
+                            invoice_line.invoice_id.timesheet_limit_date)
+                    ]])
+                    constrained_timesheets = self.env[
+                        'account.analytic.line'].search(domain)
+                    to_clean += old_timesheets - constrained_timesheets
```

Comment:
Just wondering, but wouldn't this do the same ?
```
to_clean += self.env['account.analytic.line'].search([('timesheet_invoice_id', '=', invoice_line.invoice_id.id), '|', ('date', '>', invoice_line.invoice_id.timesheet_limit_date), ('date', '=', False)])
```
Then I feel like this could be improved heavily without looping on all the created lines and doing the searches on each iteration as long as you can collect all the related invoices. Maybe add a comment and let's not spend more time on it at the moment...